### ISSUE TITLE
运行官网线性回归例子出错

### ISSUE THREAD

**Issue by LMXKO** [PaddlePaddle/Paddle#21792]
- 版本、环境信息：
    1）PaddlePaddle版本：1.6
    2）CPU：Intel(R) Xeon(R) Silver 4110 CPU @ 2.10GHz
    3）GPU：无GPU
    4）系统环境：Centos，Python3
- 模型信息
    1）线性回归 链接：https://www.paddlepaddle.org.cn/documentation/docs/zh/beginners_guide/basics/fit_a_line/README.cn.html
- 复现信息：如为报错，请给出复现环境、复现步骤
- 问题描述：
[==================================================]ing/housing.data not found, downloading http://paddlemodels.bj.bcebos.com/uci_housing/housing.data
2019-12-17 17:05:05,210-INFO: font search path ['/usr/lib64/python2.7/site-packages/matplotlib/mpl-data/fonts/ttf', '/usr/lib64/python2.7/site-packages/matplotlib/mpl-data/fonts/afm', '/usr/lib64/python2.7/site-packages/matplotlib/mpl-data/fonts/pdfcorefonts']
2019-12-17 17:05:05,548-INFO: generated new fontManager
W1217 17:05:06.042307 137919 init.cc:206] *** Aborted at 1576573506 (unix time) try "date -d @1576573506" if you are using GNU date ***
W1217 17:05:06.043684 137919 init.cc:206] PC: @                0x0 (unknown)
W1217 17:05:06.043839 137919 init.cc:206] *** SIGSEGV (@0x50) received by PID 137919 (TID 0x7fa1fda06740) from PID 80; stack trace: ***
W1217 17:05:06.044931 137919 init.cc:206]     @     0x7fa1fd2195e0 (unknown)
W1217 17:05:06.045966 137919 init.cc:206]     @     0x7fa1fd7fdaa6 _dl_relocate_object
W1217 17:05:06.046989 137919 init.cc:206]     @     0x7fa1fd805c9c dl_open_worker
W1217 17:05:06.048017 137919 init.cc:206]     @     0x7fa1fd801314 _dl_catch_error
W1217 17:05:06.049047 137919 init.cc:206]     @     0x7fa1fd80530b _dl_open
W1217 17:05:06.057346 137919 init.cc:206]     @     0x7fa1fc871a72 do_dlopen
W1217 17:05:06.058544 137919 init.cc:206]     @     0x7fa1fd801314 _dl_catch_error
W1217 17:05:06.059653 137919 init.cc:206]     @     0x7fa1fc871b32 __GI___libc_dlopen_mode
W1217 17:05:06.060737 137919 init.cc:206]     @     0x7fa1fc84a2a5 init
W1217 17:05:06.061769 137919 init.cc:206]     @     0x7fa1fd216e20 __GI___pthread_once
W1217 17:05:06.062904 137919 init.cc:206]     @     0x7fa1fc84a3bc __GI___backtrace
W1217 17:05:06.064676 137919 init.cc:206]     @     0x7fa1416fb665 paddle::platform::GetTraceBackString<>()
W1217 17:05:06.065652 137919 init.cc:206]     @     0x7fa1416fb9f1 paddle::platform::EnforceNotMet::EnforceNotMet()
W1217 17:05:06.066740 137919 init.cc:206]     @     0x7fa141e894bb paddle::operators::MulOp::InferShape()
W1217 17:05:06.068117 137919 init.cc:206]     @     0x7fa142ecb4ba paddle::framework::OperatorWithKernel::RunImpl()
W1217 17:05:06.070210 137919 init.cc:206]     @     0x7fa142ecbe91 paddle::framework::OperatorWithKernel::RunImpl()
W1217 17:05:06.071259 137919 init.cc:206]     @     0x7fa142ec78f0 paddle::framework::OperatorBase::Run()
W1217 17:05:06.073002 137919 init.cc:206]     @     0x7fa14189d526 paddle::framework::Executor::RunPreparedContext()
W1217 17:05:06.073885 137919 init.cc:206]     @     0x7fa1418a0cbf paddle::framework::Executor::Run()
W1217 17:05:06.074337 137919 init.cc:206]     @     0x7fa1416e607d _ZZN8pybind1112cpp_function10initializeIZN6paddle6pybindL22pybind11_init_core_avxERNS_6moduleEEUlRNS2_9framework8ExecutorERKNS6_11ProgramDescEPNS6_5ScopeEibbRKSt6vectorISsSaISsEEE102_vIS8_SB_SD_ibbSI_EINS_4nameENS_9is_methodENS_7siblingEEEEvOT_PFT0_DpT1_EDpRKT2_ENUlRNS_6detail13function_callEE1_4_FUNES10_
W1217 17:05:06.074770 137919 init.cc:206]     @     0x7fa14172ab5e pybind11::cpp_function::dispatcher()
W1217 17:05:06.075951 137919 init.cc:206]     @     0x7fa1fd50a672 PyEval_EvalFrameEx
W1217 17:05:06.077107 137919 init.cc:206]     @     0x7fa1fd50d08d PyEval_EvalCodeEx
W1217 17:05:06.078240 137919 init.cc:206]     @     0x7fa1fd50a58c PyEval_EvalFrameEx
W1217 17:05:06.079326 137919 init.cc:206]     @     0x7fa1fd50d08d PyEval_EvalCodeEx
W1217 17:05:06.080461 137919 init.cc:206]     @     0x7fa1fd50a58c PyEval_EvalFrameEx
W1217 17:05:06.081612 137919 init.cc:206]     @     0x7fa1fd50d08d PyEval_EvalCodeEx
W1217 17:05:06.082700 137919 init.cc:206]     @     0x7fa1fd50a58c PyEval_EvalFrameEx
W1217 17:05:06.083768 137919 init.cc:206]     @     0x7fa1fd50d08d PyEval_EvalCodeEx
W1217 17:05:06.084816 137919 init.cc:206]     @     0x7fa1fd50d192 PyEval_EvalCode
W1217 17:05:06.085896 137919 init.cc:206]     @     0x7fa1fd5265cf (unknown)
W1217 17:05:06.086952 137919 init.cc:206]     @     0x7fa1fd52779e PyRun_FileExFlags
Segmentation fault

**Comment by zhangting2020** [comment]
你好，我使用`pip install paddlepaddle==1.6.0`安装了paddle，下载官方book，使用`python train.py`执行fit a line这个demo的脚本，可以正常运行，未复现错误。

**Comment by LMXKO** [comment]
代码如下：
# coding=utf-8
from __future__ import print_function
import paddle
import paddle.fluid as fluid
import numpy as np
import math
import sys


BATCH_SIZE = 32

# 数据集定义
train_reader = paddle.batch(paddle.reader.shuffle(paddle.dataset.uci_housing.train(), buf_size=500), batch_size=BATCH_SIZE)
test_reader = paddle.batch(paddle.reader.shuffle(paddle.dataset.uci_housing.test(), buf_size=500), batch_size=BATCH_SIZE)

# 变量定义
x = fluid.layers.data(name="x", shape=[None, 13], dtype='float32')
y = fluid.layers.data(name="y", shape=[None, 1], dtype='float32')
y_predict = fluid.layers.fc(input=x, size=1, act=None)

# 全局初始化
main_program = fluid.default_main_program()
startup_program = fluid.default_startup_program()

# 计算损失函数
cost = fluid.layers.square_error_cost(input=y_predict, label=y)
avg_loss = fluid.layers.mean(cost)

# 定义优化器
test_program = main_program.clone(for_test=True)
sgd_op = fluid.optimizer.SGD(learning_rate=0.001)
sgd_op.minimize(avg_loss)

# 定义运算环境 CPU or GPU
use_cuda = False
place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace()
exe = fluid.Executor(place)
feeder = fluid.DataFeeder(place=place, feed_list=[x, y])

# 定义训练过程
epoch_num = 100
model_path = "../model/linear.model"


def train_test(exec, program, reader, feeder, fetch_list):
    acc = 1 * [0]
    cnt = 0
    for data_test in reader():
        out = exec.run(program=program,
                       feed=feeder.feed(data_test),
                       fetch_list=fetch_list)
        acc = [x_c[0] + x_c[1][0] for x_c in zip(acc, out)]
        cnt += 1
    return [x_d / cnt for x_d in acc]


step = 0
for i in range(epoch_num):
    for data_train in train_reader():
        avg_loss_value = exe.run(main_program,
                           feed=feeder.feed(data_train),
                           fetch_list=[avg_loss])
        if step % 10 == 0:
            print("训练第{}轮,\t损失值{}".format(step, avg_loss_value[0]))
        if step % 50 == 0:
            test_metics = train_test(exec=exe,
                                     program=test_program,
                                     reader=test_reader,
                                     fetch_list=[avg_loss_value.name],
                                     feeder=feeder)
            print("测试集损失值{}".format(step, test_metics[0]))
            if test_metics[0] < 2:
                break
        step += 1
        if math.isnan(float(avg_loss_value[0])):
            sys.exit("损失NaN， 训练失败")
        if model_path is not None:
            fluid.io.save_inference_model(model_path, ["x"], [y_predict], exe)

# 定义测试过程
# with fluid.scope_guard(fluid.core.Scope()):
#     [program, feed_names, fetch_targets] = fluid.io.load_inference_model(model_path, exe)
#     reader = paddle.batch(paddle.dataset.uci_housing.test(), batch_size=BATCH_SIZE)
#     data = next(reader())
#     feature = np.array([d[0] for d in data]).astype("float32")
#     label = np.array([d[1] for d in data]).astype("float32")
#     assert feed_names == "x"
#     res = exe.run(program,
#                   feed={feed_names[0]: np.array(feature)},
#                   fetch_list=fetch_targets)
#     print("波士顿房价预测结果:")
#     for ids, val in enumerate(res[0]):
#         print("%d: %.2f" % (ids, val))
#     print("波士顿房价真实结果:")
#     for ids, val in enumerate(label):
#         print("%d: %.2f" % (ids, val))

**Comment by LMXKO** [comment]
代码如下：
`# coding=utf-8
from __future__ import print_function
import paddle
import paddle.fluid as fluid
import numpy as np
import math
import sys


BATCH_SIZE = 32

# 数据集定义
train_reader = paddle.batch(paddle.reader.shuffle(paddle.dataset.uci_housing.train(), buf_size=500), batch_size=BATCH_SIZE)
test_reader = paddle.batch(paddle.reader.shuffle(paddle.dataset.uci_housing.test(), buf_size=500), batch_size=BATCH_SIZE)

# 变量定义
x = fluid.layers.data(name="x", shape=[None, 13], dtype='float32')
y = fluid.layers.data(name="y", shape=[None, 1], dtype='float32')
y_predict = fluid.layers.fc(input=x, size=1, act=None)

# 全局初始化
main_program = fluid.default_main_program()
startup_program = fluid.default_startup_program()

# 计算损失函数
cost = fluid.layers.square_error_cost(input=y_predict, label=y)
avg_loss = fluid.layers.mean(cost)

# 定义优化器
test_program = main_program.clone(for_test=True)
sgd_op = fluid.optimizer.SGD(learning_rate=0.001)
sgd_op.minimize(avg_loss)

# 定义运算环境 CPU or GPU
use_cuda = False
place = fluid.CUDAPlace(0) if use_cuda else fluid.CPUPlace()
exe = fluid.Executor(place)
feeder = fluid.DataFeeder(place=place, feed_list=[x, y])

# 定义训练过程
epoch_num = 100
model_path = "../model/linear.model"


def train_test(exec, program, reader, feeder, fetch_list):
    acc = 1 * [0]
    cnt = 0
    for data_test in reader():
        out = exec.run(program=program,
                       feed=feeder.feed(data_test),
                       fetch_list=fetch_list)
        acc = [x_c[0] + x_c[1][0] for x_c in zip(acc, out)]
        cnt += 1
    return [x_d / cnt for x_d in acc]


step = 0
for i in range(epoch_num):
    for data_train in train_reader():
        avg_loss_value = exe.run(main_program,
                           feed=feeder.feed(data_train),
                           fetch_list=[avg_loss])
        if step % 10 == 0:
            print("训练第{}轮,\t损失值{}".format(step, avg_loss_value[0]))
        if step % 50 == 0:
            test_metics = train_test(exec=exe,
                                     program=test_program,
                                     reader=test_reader,
                                     fetch_list=[avg_loss_value.name],
                                     feeder=feeder)
            print("测试集损失值{}".format(step, test_metics[0]))
            if test_metics[0] < 2:
                break
        step += 1
        if math.isnan(float(avg_loss_value[0])):
            sys.exit("损失NaN， 训练失败")
        if model_path is not None:
            fluid.io.save_inference_model(model_path, ["x"], [y_predict], exe)

# 定义测试过程
# with fluid.scope_guard(fluid.core.Scope()):
#     [program, feed_names, fetch_targets] = fluid.io.load_inference_model(model_path, exe)
#     reader = paddle.batch(paddle.dataset.uci_housing.test(), batch_size=BATCH_SIZE)
#     data = next(reader())
#     feature = np.array([d[0] for d in data]).astype("float32")
#     label = np.array([d[1] for d in data]).astype("float32")
#     assert feed_names == "x"
#     res = exe.run(program,
#                   feed={feed_names[0]: np.array(feature)},
#                   fetch_list=fetch_targets)
#     print("波士顿房价预测结果:")
#     for ids, val in enumerate(res[0]):
#         print("%d: %.2f" % (ids, val))
#     print("波士顿房价真实结果:")
#     for ids, val in enumerate(label):
#         print("%d: %.2f" % (ids, val))`

**Comment by zhangting2020** [comment]
我试着调整了下你的代码，需要在for i in range(epoch_num):之前加一句
```python
exe.run(startup_program)
```
这样应该可以正常运行。

**Comment by zhangting2020** [comment]
建议使用下面链接中的fit_a_line教程代码
教程的代码：https://github.com/PaddlePaddle/book/tree/develop/01.fit_a_line
